File: api/python/top_mail_ru.py
import struct
import urllib
import urllib2
import json
import hashlib

class TopMailRu(object):

  apikey = ""
  session = ""

  def __init__(self, apikey=''):
    self.apikey = apikey	

  def request(self, path, args):
    url = 'https://top.mail.ru' + path + '?' + urllib.urlencode(args)
    try:
      resp = urllib2.urlopen(url)
      jr = json.loads(resp.read())
    except IOError, e:
      if hasattr(e, 'reason'):
        print 'We failed to reach a server.'
        print 'Reason: ', e.reason
      elif hasattr(e, 'code'):
        print 'The server couldn\'t fulfill the request.'
        print 'Error code: ', e.code
    else:
      return jr

  def registerSite(self, args):
    args['apikey'] = self.apikey
    args['session'] = self.session
    return self.request('/json/add', args)

  def editSite(self, id, password, args):
    args['apikey'] = self.apikey
    args['session'] = self.session
    args['id'] = id
    args['password'] = password
    return self.request('/json/edit', args)

  def getCode(self, id, password, args):
    args['apikey'] = self.apikey
    args['session'] = self.session
    args['id'] = id
    args['password'] = password
    return self.request('/json/code', args)

  def getStat(self, id, password, type, args):
    args['apikey'] = self.apikey
    args['session'] = self.session
    args['id'] = id
    args['password'] = password
    return self.request('/json/'+type, args)

  def setSession(self, session):
    self.session = session

  def login(self, id, password):
    args = { 'id': id, 'password': password, 'action': 'json' }
    args['apikey'] = self.apikey
    args['session'] = self.session
    res = self.request('/json/login', args)
    if res.get('session'):
      self.session = res['session']
    return res.get('logged') == 'yes'

File: api/php/top_mail_ru.php
<?php

/**
 * Class TopMailRu
 */
class TopMailRu
{
    const BASE_URL = 'http://top.mail.ru';

    private $apiKey;
    private $returnAsArray;
    private $session;

    /**
     * TopMailRu constructor.
     * @param $apiKey
     * @param $returnAsArray
     */
    public function __construct($apiKey, $returnAsArray)
    {
        $this->apiKey = $apiKey;
        $this->returnAsArray = $returnAsArray;
    }

    /**
     * @param $o1
     * @param $o2
     * @return string
     */
    protected function bitxor($o1, $o2)
    {
        $result = '';
        $runs = strlen($o1);
        for ($i = 0; $i < $runs; $i++)
            $result .= $o1[$i] ^ $o2[$i];
        return $result;
    }

    /**
     * @param $path
     * @param $argsArray
     * @param $returnAsArray
     * @return mixed|null
     */
    protected function request($path, $argsArray, $returnAsArray)
    {
        $url = static::BASE_URL . $path . '?' . http_build_query($argsArray);
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        $data = null;
        try {
            $data = curl_exec($ch);
            curl_close($ch);
            $data = json_decode($data, $returnAsArray);
        } catch (Exception $e) {
            echo "Exception: {$e->getCode()} ({$e->getMessage()})", PHP_EOL;
        }

        return $data;
    }

    /**
     * @return array
     */
    protected function getKeyAndSession()
    {
        $args = array();
        if ($this->apiKey) {
            $args['apikey'] = $this->apiKey;
        }
        if ($this->session) {
            $args['session'] = $this->session;
        }
        return $args;
    }

    /**
     * @param $args
     * @return mixed|null
     */
    public function registerSite($args)
    {
        $args += $this->getKeyAndSession();
        return $this->request('/json/add', $args, $this->returnAsArray);
    }

    /**
     * @param $id
     * @param $password
     * @param $args
     * @return mixed|null
     */
    public function editSite($id, $password, $args)
    {
        $args += $this->getKeyAndSession();
        $args['id'] = $id;
        $args['password'] = $password;
        return $this->request('/json/edit', $args, $this->returnAsArray);
    }

    /**
     * @param $id
     * @param $password
     * @param $args
     * @return mixed|null
     */
    public function getCode($id, $password, $args)
    {
        $args += $this->getKeyAndSession();
        $args['id'] = $id;
        $args['password'] = $password;
        return $this->request('/json/code', $args, $this->returnAsArray);
    }

    /**
     * @param $session
     */
    public function setSession($session)
    {
        $this->session = $session;
    }

    /**
     * @param $id
     * @param $password
     * @return bool
     */
    public function login($id, $password)
    {
        $args = array_merge(
            array('id' => $id, 'password' => $password, 'action' => 'json'),
            $this->getKeyAndSession()
        );
        $res = $this->request('/json/login', $args, true);
        if ($res['session']) {
            $this->session = $res['session'];
        }
        return isset($res['logged']) && $res['logged'] === 'yes';
    }

    /**
     * @param $id
     * @param $ph
     * @return bool
     */
    public function loginByHash($id, $ph)
    {
        $args = array_merge(
            array('id' => $id, 'action' => 'json'),
            $this->getKeyAndSession()
        );
        $response = $this->request('/json/login', $args, true);
        if ($response['session']) {
            $this->session = $response['session'];
        }
        if (isset($response['salt'], $response['seed'])) {
            $hash2 = sha1($response['salt'] . $ph, true);
            $hash3 = $this->bitxor($ph, sha1($response['seed'] . $hash2, true));
            $args = array_merge(
                array('id' => $id, 'action' => 'json', 'ph' => bin2hex($hash3), 'seed' => $response['seed']),
                $this->getKeyAndSession()
            );
            $response = $this->request('/json/login', $args, true);
        }

        return (isset($response['logged']) && $response['logged'] === 'yes');
    }

    /**
     * @param $id
     * @param $password
     * @param $type
     * @param $args
     * @return mixed|null
     */
    public function getStat($id, $password, $type, $args)
    {
        $args += $this->getKeyAndSession();
        $args['id'] = $id;
        $args['password'] = $password;
        return $this->request('/json/' . $type, $args, $this->returnAsArray);
    }
}
